models/firefly.py

```
@@ -14,35 +14,103 @@
+    greater than the agent's blinking frequency.
+    """
+    # Calculate the blink parameter
+    blink_frequency = agent.get_attr(BLINK_FREQUENCY)
+    time_since_last_blink = agent.get_attr(LAST_BLINKED_AT) - agent.duration
+
+    # Get the previous group name
+    old_group = agent.group_name()
+
+    # Turn OFF if the firefly is ON
+    if old_group == FIREFLY_ON:
+        agent.set_prim_group(FIREFLY_OFF)
+
+    # Turn ON if the blinking time has arrived
+    elif time_since_last_blink >= blink_frequency:
+        agent.set_prim_group(FIREFLY_ON)
+        # Reset the attribute
+        agent.set_attr(LAST_BLINKED_AT, agent.duration)
+
+    # Perform the actual switch
+    get_model(agent.exec_key).add_switch(
+        str(agent), old_group, agent.group_name()
+    )
+
+
+def adjust_blink_frequency(agent, **kwargs):
+    """
+    Inreases or decreases the agent's blinking frequency based on the average
+    of its neighbors. If the blinking frequency is not initialized for this
+    agent, it assigns a random frequency value within the specified range.
+    """
+    # Initialize the blink frequency if not initialized before
+    if agent.get_attr(BLINK_FREQUENCY) is None:
+        frequency = random.randint(
+            DEF_MIN_BLINK_FREQUENCY, DEF_MAX_BLINK_FREQUENCY
+        )
+        time = agent.duration
+        agent.set_attr(BLINK_FREQUENCY, frequency)
+        agent.set_attr(LAST_BLINKED_AT, time)
+
+        if DEBUG.debug:
+            print(f"Set {agent}'s blink frequency to {frequency}")
+            print(f"Set {agent}'s last blinked at time to {time}")
+
+    # Get the average blinking frequency of the neighbours
+    else:
+        pass
 
 
 def firefly_action(agent, **kwargs):
     """
     A simple default agent action.
     """
-    if DEBUG.debug:
-        print("Agent {} is located at {}".format(agent.name, agent.get_pos()))
+    model = get_model(agent.exec_key)
+    if model is None:
+        print("ERROR: get_model() returned None.")
```

Maybe throw here instead of printing?
get_model should not return None inside agent action.

---

I saw this thing in the forest fire model and thought it is needed: https://github.com/TandonDevOps/IndraABM/blob/master/models/forest_fire.py#L66

Sure, I can change it